kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLogManager.java

```diff
@@ -60,62 +60,43 @@ public PartitionLogManager(KafkaServiceConfiguration kafkaConfig,
                                OrderedExecutor recoveryExecutor) {
         this.kafkaConfig = kafkaConfig;
         this.requestStats = requestStats;
-        this.logMap = Maps.newConcurrentMap();
         this.entryFilters = entryFilters;
         this.time = time;
         this.kafkaTopicLookupService = kafkaTopicLookupService;
         this.producerStateManagerSnapshotBuffer = producerStateManagerSnapshotBuffer;
         this.recoveryExecutor = recoveryExecutor;
     }
 
-    public PartitionLog getLog(TopicPartition topicPartition, String namespacePrefix) {
+    public CompletableFuture<PartitionLog> getLog(TopicPartition topicPartition, String namespacePrefix) {
         String kopTopic = KopTopic.toString(topicPartition, namespacePrefix);
         String tenant = TopicName.get(kopTopic).getTenant();
         ProducerStateManagerSnapshotBuffer prodPerTenant = producerStateManagerSnapshotBuffer.apply(tenant);
-        PartitionLog res =  logMap.computeIfAbsent(kopTopic, key -> {
-            PartitionLog partitionLog = new PartitionLog(kafkaConfig, requestStats,
-                    time, topicPartition, key, entryFilters,
-                    kafkaTopicLookupService,
-                    prodPerTenant, recoveryExecutor);
-
-            CompletableFuture<PartitionLog> initialiseResult = partitionLog
-                    .initialise();
-
-            initialiseResult.whenComplete((___, error) -> {
-                if (error != null) {
-                    // in case of failure we have to remove the CompletableFuture from the map
-                    log.error("Failed to recovery of {}", key, error);
-                    logMap.remove(key, partitionLog);
-                }
-            });
-
-            return partitionLog;
-        });
-        if (res.isInitialisationFailed()) {
-            log.error("Failed to initialize of {}", kopTopic);
-            logMap.remove(kopTopic, res);
+        CompletableFuture<PartitionLog> future = logMap.computeIfAbsent(kopTopic, key -> new PartitionLog(
+                kafkaConfig, requestStats, time, topicPartition, key, entryFilters, kafkaTopicLookupService,
+                prodPerTenant, recoveryExecutor
+        ).initialise());
+        // We need to wait here to avoid the message disorder
+        try {
+            future.get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException e) {
+            log.error("Failed to get PartitionLog for {} under {}", topicPartition, namespacePrefix, e.getCause());
```

Comment:
Do we need to remove the failed future?

Reply:
It will be removed in the `exceptionally` callback of the future returned.

Reply:
This PR has a serious problem so that it's draft now.